### ISSUE TITLE
Maven Help Plugin is not able to browse the repository

### ISSUE THREAD

**Issue by dzikoysk** [dzikoysk/reposilite#476]
> Hi there,
> I am encountering a problem that *may* be related to this issue.
> I am writing a shell script that is trying to determine the current version of the POM by running

```
mvn help:evaluate -Dexpression=project.version
```
> I am getting the following error from maven:
```
> No plugin found for prefix 'help' in the current project and in the plugin groups [org.apache.maven.plugins, org.codehaus.mojo] available from the repositories
```
>I am trying hard to figure out what the actual problem might be. I do not entirely understand what is happening, but it seems, > that maven is trying to download 
> `https://<my reposilte mirror>/org/apache/maven/plugins/maven-metadata.xml` to figure out what plugins are available for the > `help` prefix? It then gets confused by what it finds - the content looks somewhat weird; it has actual plugin-names at places > where I think should be version strings -- that's what brought me here, actually. So, reposilite generates the content of this file on the fly?  

> I am sorry, I do not understanding half of it. Do you think my problem could be related?

-- EDIT

> Just a quick follow-up: When copy the contents from the original repository over to my local repository (overwriting the cached > version in $HOME/.m2/repositories/org/apache/maven/plugins/maven-metadata-<my reposilite mirror>.xml, the mvn > help:evaluate command executes just fine. So there must be some relation, no?

_Originally posted by @ldegen in https://github.com/dzikoysk/reposilite/issues/461#issuecomment-832790409_

**Comment by dzikoysk** [comment]
Associated project:

* https://github.com/apache/maven-help-plugin

**Comment by dzikoysk** [comment]
@ldegen could add more context about the issue? Let's start with log during the execution of your help command. (You should be able to find full details about the requests in log.txt file)

**Comment by ldegen** [comment]
@dzikosk thanks for the quick reply.

Ok, let me see. To avoid any confusion due to cached artifacts,
I started by deleting the whole `/org/apache/maven/plugins` from
both my local repository and the reposilite mirror.

I also cleared the reposilite log file `log.txt`.
I am running version 2.9.15 btw.

Next, I run

```
mvn -X help:evaluate -Dexpression=project.version > maven-output.txt 2>&1
```

I created a gist for the two output files: https://gist.github.com/ldegen/00a61bc84834d2578a9c6ec23cfd839e

I am running this on a trivial maven test project. Nothing much in there, really, but if you want, I think I could upload it to github.

---

In the meantime I found this: https://maven.apache.org/guides/introduction/introduction-to-plugin-prefix-mapping.html
So, if I understand this correctly, my problem is probably not specific to the maven-help-plugin, but simply about the way maven discovers "mojos" when running plugin executions directly from the cli.  
It is says there that
> By default, Maven will search the groupId org.apache.maven.plugins for prefix-to-artifactId mappings for the plugins it needs to perform a given build.

And if you look for instance at https://repo1.maven.org/maven2/org/apache/maven/plugins/maven-metadata.xml, you see that there are a ton of `<plugin>`-Elements listed, which looks like it is just how maven figures out what plugins it should use for which prefix.

I had a quick look at the reposilite source code. Am I correct to assume that this class https://github.com/dzikoysk/reposilite/blob/master/reposilite-backend/src/main/java/org/panda_lang/reposilite/metadata/Metadata.java is a representation of the XML-Content generated for those metadata.xml-files? If yes, shouldn't there be some `plugins`-Member as well?
That is all assuming that those metadata-files are actually used for what I think they are -- I really do not know very much about how maven repositories work.

**Comment by ldegen** [comment]
It's strange, though: even when I run maven with the fully-qualified plugin name, i.e.
```
mvn -X org.apache.maven.plugins:maven-help-plugin:3.2.0:evaluate -Dexpression=project.version
```
I still get the same error. Which makes no sense to me, since in this case maven does not have to look up any prefix. Very strange, indeed.

**Comment by dzikoysk** [comment]
Thanks for details. Well, Reposilite doesn't support plugins at this moment, as you noticed in Metadata class representation. I'd need to discover this topic as I don't know much about it. 

> I am running this on a trivial maven test project. Nothing much in there, really, but if you want, I think I could upload it to 
github.

If you could provide a template project that I could use during tests I'd be glad, it'll just simplify the process.

> 
> 
> It's strange, though: even when I run maven with the fully-qualified plugin name, i.e.
> 
> ```
> mvn -X org.apache.maven.plugins:maven-help-plugin:3.2.0:evaluate -Dexpression=project.version
> ```
> 
> I still get the same error. Which makes no sense to me, since in this case maven does not have to look up any prefix. Very strange, indeed.

It's hard for me to say, but there is a high chance that it requests everything even if it's not really required.

**Comment by ldegen** [comment]
>>   I am running this on a trivial maven test project. Nothing much in there, really, but if you want, I think I could upload it to github.

> If you could provide a template project that I could use during tests I'd be glad, it'll just simplify the process.

Sure, here you go: https://github.com/ldegen/repositlite-issue-476

**Comment by dzikoysk** [comment]
Thanks, I'll try to investigate the issue in a next few days.

**Comment by dzikoysk** [comment]
Tbh I don't know when this plugin decides to request remote repository, but even if I forced download from my remote repository it works quite fine I think.

```
$ mvn org.apache.maven.plugins:maven-help-plugin:3.1.1:evaluate -Dexpression=project.version

[INFO] Scanning for projects...
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Build Order:
[INFO] 
[INFO] reposilite-parent                                                  [pom]
[INFO] reposilite-backend                                                 [jar]
Downloading from panda-repository: https://repo.panda-lang.org/releases/org/apache/maven/plugins/maven-help-plugin/3.1.1/maven-help-plugin-3.1.1.pom
Downloaded from panda-repository: https://repo.panda-lang.org/releases/org/apache/maven/plugins/maven-help-plugin/3.1.1/maven-help-plugin-3.1.1.pom (9.4 kB at 7.4 kB/s)
Downloading from panda-repository: https://repo.panda-lang.org/releases/org/apache/maven/plugins/maven-help-plugin/3.1.1/maven-help-plugin-3.1.1.jar
Downloaded from panda-repository: https://repo.panda-lang.org/releases/org/apache/maven/plugins/maven-help-plugin/3.1.1/maven-help-plugin-3.1.1.jar (64 kB at 180 kB/s)
[INFO] 
[INFO] ------------------< org.panda-lang:reposilite-parent >------------------
[INFO] Building reposilite-parent 2.9.22-SNAPSHOT                         [1/2]
[INFO] --------------------------------[ pom ]---------------------------------
[INFO] 
[INFO] --- maven-help-plugin:3.1.1:evaluate (default-cli) @ reposilite-parent ---
[INFO] No artifact parameter specified, using 'org.panda-lang:reposilite-parent:pom:2.9.22-SNAPSHOT' as project.
[INFO] 
2.9.22-SNAPSHOT
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for reposilite-parent 2.9.22-SNAPSHOT:
[INFO] 
[INFO] reposilite-parent .................................. SUCCESS [  2.620 s]
[INFO] reposilite-backend ................................. SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  5.363 s
[INFO] Finished at: 2021-05-07T13:43:54+02:00
[INFO] ------------------------------------------------------------------------
```

I think that during the generating of metadata file we can just verify by artifact name if it may be a plugin and just move it to the `plugins` section. Unfortunately I don't think I'm able to implement this before Reposilite 3.0. #387 directly modifies files associated with metadata files. As a workaround you can use this route:

* https://reposilite.com/docs/indexing#latest-info

**Comment by ldegen** [comment]
There is a misunderstanding, I think. This has nothing to do with the maven-help-plugin, or any other specific plugin.

The problem occurs when the maven cli tries to lookup  plugin *metadata*. E.g. when you run something like `mvn foo:bar`, maven needs to first figure out which plugin provides the `foo`-prefix. It does so by iterating through a list of well-known groupIds. These can be configured, but the hard-wired fall back is always`org.apache.maven.plugins`, I think.
For each of these groupIds, maven needs the contents of the associated `maven-metadata.xml` file. Within this file, it looks for a `<plugins>` which in turn contains the prefix-to-plugin mappings.

Now, if maven thinks, that the version of that metadata within your local cache is "up-to-date", it will not query the remote repository.
You can check this in your local repo, for me, it's in `~/.m2/repository/org/apache/maven/plugins`. There you should see one or more files named `maven-metadata-<some repository id>.xml`, which is the cached version for one particular remote repository.
If you delete those files (I went ahead and wiped the whole directory, actually), you force maven to download all the metadata again the next time it needs to lookup a plugin prefix.
And this is what I did to trigger the problem: The metadata file generated by reposilite simply does not contain the required information.

> I think that during the generating of metadata file we can just verify by artifact name if it may be a plugin and just move it to the plugins section.
Here, you lost me, I'm afraid. We don't have an artifact name to begin with, or do we?

I just had another look at the MetadataService, trying to understand how it works. To tell the truth, it's probably much to late , and I should get some sleep first :-)
But I was wondering: is the metadata only generated from the artifacts that are actually locally stored/cached?

**Comment by dzikoysk** [comment]
Yeah, probably, I'm so tired this week :( I've just checked what does this goal do and how it works. The issue is marked as enhancement because metadata file in facts misses `plugins` section.

I'm not really into maven cli despite of some basic goals used during building some projects. My approach to reproduce is:

1. Add some custom plugin repository through:
```xml
    <pluginRepositories>
        <pluginRepository>
            <id>panda-repository</id>
            <url>https://repo.panda-lang.org/releases</url>
        </pluginRepository>
    </pluginRepositories>
```
2. Delete `~\.m2\repository\org\apache\maven\plugins`
3. Run the given command

Wiping the plugins directory does not really change anything in my case, e.g.:
```
Downloaded from panda-repository: https://repo.panda-lang.org/releases/org/apache/maven/plugins/maven-assembly-plugin/2.2-beta-5/maven-assembly-plugin-2.2-beta-5.pom (15 kB at 59 kB/s)
Downloading from panda-repository: https://repo.panda-lang.org/releases/org/apache/maven/plugins/maven-plugins/16/maven-plugins-16.pom
Downloaded from panda-repository: https://repo.panda-lang.org/releases/org/apache/maven/plugins/maven-plugins/16/maven-plugins-16.pom (13 kB at 63 kB/s)
```

But I see that Maven requests the plugin's metadata file used in command at the end of the downloads:

```
Downloading from panda-repository: https://repo.panda-lang.org/releases/org/apache/maven/plugins/maven-metadata.xml
Downloading from central: https://repo.maven.apache.org/maven2/org/apache/maven/plugins/maven-metadata.xml
Downloading from panda-repository: https://repo.panda-lang.org/releases/org/codehaus/mojo/maven-metadata.xml
Downloaded from panda-repository: https://repo.panda-lang.org/releases/org/codehaus/mojo/maven-metadata.xml (603 B at 4.0 kB/s)
Downloaded from panda-repository: https://repo.panda-lang.org/releases/org/apache/maven/plugins/maven-metadata.xml (1.0 kB at 3.1 kB/s)
Downloaded from central: https://repo.maven.apache.org/maven2/org/apache/maven/plugins/maven-metadata.xml (14 kB at 41 kB/s)
Downloading from panda-repository: https://repo.panda-lang.org/releases/org/apache/maven/plugins/maven-help-plugin/maven-metadata.xml
Downloading from central: https://repo.maven.apache.org/maven2/org/apache/maven/plugins/maven-help-plugin/maven-metadata.xml
Downloaded from central: https://repo.maven.apache.org/maven2/org/apache/maven/plugins/maven-help-plugin/maven-metadata.xml (652 B at 6.8 kB/s)
Downloaded from panda-repository: https://repo.panda-lang.org/releases/org/apache/maven/plugins/maven-help-plugin/maven-metadata.xml (263 B at 1.5 kB/s)
[INFO] 
```

I agree that the metadata content is incorrect, I just couldn't force the error. Speaking of MetadataService, atm it generates metadata content during the first request to the Reposilite instance and keeps it in memory (also updates the version stored on disk). Data stored in metadata file bases on the current repository content to avoid various desync issues.

**Comment by ldegen** [comment]
>atm it generates metadata content during the first request to the Reposilite instance and keeps it in memory (also updates the version stored on disk). Data stored in metadata file bases on the current repository content to avoid various desync issues.

I see two potential problems with this approach:
- You say the metadata is only created once (upon the first request). Is it ever updated when the contents of the (local) repository change?
- This also means that Reposilite may not "see" certain plugins because they were never requested and thus are not cached, right?

Both problems are probably out of scope for this issue. I don't think they are actually affecting my current situation (much), so let's ignore that for now.

The remaining problem then would be generating the correct metadata for all artifacts that the reposilite instance actually does know about. Regarding your earlier remark on somehow using the artifact name (id?) to decide whether it is a plugin: It seems that there actually *is* a naming convention, which is described in this paragraph: https://maven.apache.org/guides/introduction/introduction-to-plugin-prefix-mapping.html#specifying-a-plugins-prefix
Strictly speaking, this is not enough to correctly detect a plugin. As far as I understand, this convention is only used for mapping the prefix - i.e. call your plugin `foo-maven-plugin`, and the maven repository will automatically map the prefix `foo` to your plugin. It might be good enough, though. 

A more fail-safe way to detect if something is a plugin would be to look into the POM and examine the contents of the `<packaging>`-Element. For plugins, it should be `maven-plugin`. 
For extra points, we would also check if the POM contains a `<plugin>`-Element for the `maven-plugin-plugin`, and if it does, check it's configuration for any`<goalPrefix>`-element to determine if the plugin explicitly overrides the prefix mapping.

I could try and put together a pull request, if you don't mind?

**Comment by dzikoysk** [comment]
By saying Reposilite instance, I mean that metadata files are regenerated every time you restart the application. Also, during the deploy, associated metadata file is purged from cache, which means it also will be regenerated. The only desync possibility is when you upload some artifacts directly to the repository e.g. through ftp and without using `purge` command.

> I think that during the generating of metadata file we can just verify by artifact name if it may be a plugin and just move it to the plugins section.

Of course, I was referring the [naming convection](https://maven.apache.org/guides/introduction/introduction-to-plugin-prefix-mapping.html#specifying-a-plugins-prefix). I like your idea for fail-safe approach to not only filter names as I proposed,  but also check pom files to verify the packaging.

If possible, please keep in mind we're before various breaking changes to Reposilite 3.0. Especially this one:

* https://github.com/dzikoysk/reposilite/pull/387

Modify a lot of fs related stuff - also metadata files. The best way would be probably moving the sources associated to this issue to new classes like `MetadataPluginXXX` to avoid unpleasant merge conflicts between PRs, master and 3.0 branch.

**Comment by ldegen** [comment]
Just a quick heads up: After digging deeper into what turned out to be quite a rabbit hole, I am now trying to create a library to solve the more general "give me metadata for that pom.xml"-problem.
Started a new repository here https://gitlab.com/ldegen/maven-repo-metadata. Nothing much to look at yet, but I am working on it.
  
My plan is to then modify or replace Reposilite's current MetadataService to use that library once the library is at least partly working.

**Comment by dzikoysk** [comment]
Yeah, it might be a better approach to experiment in a new repo, due to #465. I'm kinda busy because of the schedule for [3.0](https://github.com/dzikoysk/reposilite/projects/2), so any extra help during the research is appreciated ❤️. I'd like to just notice, that in the end it should be probably rewritten as a metadata domain impl to avoid a scenerio where we're limited to the library which cannot be directly modified within this repository and has a big impact on the Maven related logic. As an example, we could use Maven as a library in Reposilite and just use some of their GAV representations, but as you could notice, we have just our own representation.

**Comment by ldegen** [comment]
On a more general node:

In the documentation for the [maven-deploy-plugin](http://maven.apache.org/plugins/maven-deploy-plugin/) it says:
> As a repository contains more than the artifacts (POMs, the metadata, MD5 and SHA1 hash files...), deploying means not only copying the artifacts, but making sure all this information is correctly updated. It's the reponsibility of the deploy plugin.

Hm. But then it is *not* the responsibility of the repository, is it?

Of course this would only work, if the repository was actually trusting the metadata that is generated by the deploy-plugin. Which, as I understand it, Reposilite does not. @dzikoysk  Could you elaborate on the rational for this decision? I remember you mentioning some "de-sync issue" somewhere, but now I cannot find it anymore. Was it in some kind of FAQ?

I will carry on with [my idea](https://gitlab.com/ldegen/maven-repo-metadata) for now, as I still think something like this  could be useful as a fallback in situation where the metadata file is missing or cannot be trusted for some reason.

**Comment by dzikoysk** [comment]
No, generating metadata file is not really a task for repository manager. In theory, repository manager should work more like a remote storage for data generated by Maven tasks. It sounds great, but irl such an approach limits the possibilities of modyfing repository content (like manual upload through ftp/dashboard, various interactions with any content located in repository). Reposilite 1.x used embedded Maven to generate various metadata files using the whole build system, but it was an overkill for huge repositories. Maven is terribly slow and the api is kinda limited to perform various stub tasks. In Reposilite 2.x I decided to reverse engineer the generator based on data located in repository and it works pretty good, despite of some lacks of the current implementation. In Reposilite 3.x I'd like to simplify the generator (2.x code base kinda scares me) and somehow mix two of these 2 approaches.

**Comment by ldegen** [comment]
>It sounds great, but irl such an approach limits the possibilities of modyfing repository content (like manual upload through ftp/dashboard, various interactions with any content located in repository, etc.).

yap. I totally agree.

**Comment by dzikoysk** [comment]
Addressed by #550
